### DISCUSSION TITLE
PDF sketches using imported or module mode?

### DISCUSSION THREAD

**Discussion by tabreturn** [py5coding/py5generator#43] [Q&A] [answered]
I'm writing my code in Thonny. How do I create PDF sketches?

**Comment by hx2A** [comment, marked as answer]
I just tested this using run_sketch:

```
def setup():
    size(400, 400, PDF, "/tmp/test.pdf")

    for _ in range(20):
        fill(255 * random(), 255 * random(), 255 * random())
        ellipse(width * random(), height * random(), 20, 20)
    exit_sketch()

```

I don't think it would work without the `exit_sketch()`.

I saw your Axidraw posts on Twitter and know PDFs can be used with that plotter. Out of curiosity, is your question related to that? Or are you making something for the Processing Community Catalog (for which PDFs are the required format and submissions are due at the end of the month)?

It happens I bought an Axidraw and it will arrive soon. I'd be happy to collaborate with you on ehancing py5's support for Axidraw.

I also need to get cracking on my py5 Processing Community Catalog submission.

**Reply by tabreturn** [reply]
Thanks! That works. I was using `exit()` instead of `exit_sketch()` 😳 

Yep -- this is for plotter stuff. The AxiDraw plays nice with Inkscape. My workflow is: write code; render as PDF (vector-based); import PDF into Inkscape; plot. So py5 'support' is already good enough for me. But AxiDraw has a Python API, which presents some exciting opportunities ...

I've already submitted something to the Processing Community Catalog -- a [py5 port of my Aquatics](https://github.com/tabreturn/processing.py-digital-aquatics). Like, a screenshot of several aquatics overlapping/surrounding a Thonny editor window (with some partially revealed source code).

Rad! Time to create a py5 plot 😎

**Reply by hx2A** [reply]
> Thanks! That works. I was using `exit()` instead of `exit_sketch()` 😳

Right, that is one of the differences with processing.py. In py5 I avoided all name conflicts with builtin functions. The other two that I can think of right now are `filter()` => `apply_filter()` and `map()` => `remap()`. These should be called out in the documentation. 

> Yep -- this is for plotter stuff. The AxiDraw plays nice with Inkscape. My workflow is: write code; render as PDF (vector-based); import PDF into Inkscape; plot. So py5 'support' is already good enough for me. But AxiDraw has a Python API, which presents some exciting opportunities ...

The AxiDraw is my birthday present to myself, even though my birthday isn't for another two weeks. I probably won't have that much time to really dive into using the AxiDraw though because I am currently swamped with other work responsibilities. My initial workflow will probably be the same as what you are doing now.

> I've already submitted something to the Processing Community Catalog -- a [py5 port of my Aquatics](https://github.com/tabreturn/processing.py-digital-aquatics). Like, a screenshot of several aquatics overlapping/surrounding a Thonny editor window (with some partially revealed source code).

Sounds great! I can't wait to see that!

> Rad! Time to create a py5 plot 😎

Post the results to twitter and I will retweet it!

**Comment by hx2A** [comment]
This also works with `run_sketch`:

```
size(400, 400, PDF, "/tmp/test2.pdf")

for _ in range(20):
    fill(255 * random(), 255 * random(), 255 * random())
    ellipse(width * random(), height * random(), 20, 20)
exit_sketch()
```

Analogous code using the SVG renderer should work as well.

BTW, after the next release, py5bot will support SVGs.
